### ISSUE TITLE
PayPal Insights Integration

### ISSUE THREAD

**Issue by angelleye** [angelleye/paypal-woocommerce#904]
We talked about doing this as a stand-alone WP plugin, but PayPal has asked us to simply include it in this plugin for now.

**Comment by angelleye** [comment]
@kcppdevelopers I'd like to add PayPal Insights as a new tab under Settings -> PayPal for WooCommerce -> General.  We can drop in the JS for the activate/modify button into this settings panel, and then we'll drop the regular JS for page content so that the PayPal banner displays based on the Insights configuration.

**Comment by angelleye** [comment]
@kcppdevelopers I'm a little bit confused here.

In our settings panel I see a check box to enable Insights, but I don't see the actual Activate / Modify button showing up anywhere..??

Also, I'm not sure why we have to fill out a "Merchant's website URL" field..??  I don't remember seeing the need for that in the Insights docs, and even if we do need it, couldn't we just use the WP site URL?

**Comment by angelleye** [comment]
@kcppdevelopers Here's a sample PayPal Insights (MUSE) plugin for Woo that PayPal put together:  https://github.com/trainerbill/woocommerce-paypal-marketing

I didn't see that sample until now.  We should review this together and probably do ours similar.  I'm still not sure what you did here because I don't see the Activate button.

**Comment by angelleye** [comment]
@kcppdevelopers Their plugin example on GitHub uses the user's API credentials.  We basically need to replicate with their plugin is doing in ours.

I originally said to do this under the Settings -> PayPal for WooCommerce -> General tab.  Instead, we need to move this into the Express Checkout settings, because it really only applies to that gateway.

In his plugin you have to click the box to enable, and then it displays the Activate / Modify button.  I feel like that's an necessary step and I'm not sure why he's doing that.  We should be able to simply display the Activate / Modify button, and use the cid if we have it available.  Without a cid the button will say Activate, and with one it would say Modify.  So of course we need to save their cid locally once we've obtained it for the user.

**Comment by kcppdevelopers** [comment]
@angelleye I am done with PayPal Insights and push to dev branch for your review.

**Comment by angelleye** [comment]
@kcppdevelopers I forgot to add one more thing to this.  We need to include an admin notice for this update.

"PayPal Insights now available in Express Checkout!  Make sure to activate Insights for valuable analytics about your visitors and increased conversion rates on your site!"

The words "activate Insights" should be a link directly to the Insights section of the EC settings panel.

**Comment by angelleye** [comment]
@kcppdevelopers  I guess they have this integrated into their Express Checkout plugin:  https://github.com/trainerbill/woocommerce-gateway-paypal-express-checkout

That's where it was done with API credentials so that they don't have to login.  We should install their EC plugin and review the code to see about doing this here too please.

**Comment by angelleye** [comment]
New official name for this is "PayPal Marketing Solutions" instead of PayPal insights.  Also, the integration in their EC plugin has some Terms and Conditions we need to include in ours as well.

![image](https://user-images.githubusercontent.com/629241/31835477-4b1d7fb4-b597-11e7-94d1-c428d8b76143.png)

**Comment by angelleye** [comment]
@kcppdevelopers I'm still not seeing API credentials included so that we don't have to sign in to activate insights..??  

Did you install the EC plugin that they have and check it out?  They are adamant that it uses API credentials so you don't have to sign in, and that's what they're asking us for here.  When I run a test, though, it's still bringing up a login window..??

**Comment by angelleye** [comment]
@kcppdevelopers It's also still making me type in my website address.  Can't we pass the {site_url} so that it gets pre-filled for us?

**Comment by angelleye** [comment]
@kcppdevelopers I think maybe I haven't been clear on what we need to do here.  Let me try to explain again what's going on here.

Woo has their own PayPal Express plugin available in the WP directory:  https://wordpress.org/plugins/woocommerce-gateway-paypal-express-checkout/

One of the PayPal reps forked that repo to his own, and he built in the PayPal Insights functionality:  https://github.com/trainerbill/woocommerce-gateway-paypal-express-checkout

When running that forked version of the plugin, we see this inside the Express Checkout Settings:

![image](https://user-images.githubusercontent.com/629241/31913537-0347819e-b80d-11e7-99e8-6650d47f3415.png)

We need ours to look and function like this, except we should try to improve it where possible.  For example, if we can pass the site_url to pre-fill the dialogs that would be good.

**Comment by angelleye** [comment]
@kcppdevelopers I installed their forked copy of the plugin, and I see it's still asking me to login when I click to Activate / Manage Insights, so I've asked them for clarification on how that is supposed to be working.

We just need to go ahead and add this additional marketing stuff along with the terms and conditions you already added the same they've done.  Notice it adjusts when you click Enable...

**Comment by angelleye** [comment]
@kcppdevelopers Okay, so they actually have 2 separate GitHub repos, one is showing Insights built into the EC plugin, and the other is for a separate stand-alone plugin:

> 1.      Example code and video for your existing PayPal Woo plugin that has EC tab accepting merchant’s PP credentials and bypasses the login to PayPal for activating MUSE (Marketing Solutions). It does require login to PayPal to ‘Manage Settings’ after activated.
> ·        GitHub: https://github.com/trainerbill/woocommerce-gateway-paypal-express-checkout
> ·        Video: https://youtu.be/tsSHL8NarAE
>  
> 2.      Example code and video for new standalone plugin that requires login to PayPal for activating MUSE (Marketing Solutions). It also requires login to PayPal to ‘Manage Settings’ after activated.
> ·        GitHub: https://github.com/trainerbill/woocommerce-paypal-marketing
> ·        Video: https://youtu.be/aGeWSQrF1MU

The video does show it skipping the login, but as you mentioned previously that could be because he's already signed in to the PayPal account.  We just need to take a closer look at the second example here to make sure we haven't missed something.

**Comment by angelleye** [comment]
@kcppdevelopers Okay, now I have verified with PayPal that the login is NOT going to be skipped.  So at this point we just need to add in the extra marketing verbiage / graphics like they have in theirs.

**Comment by angelleye** [comment]
@kcppdevelopers We also need to make sure we have a good admin notice that notifies the user about PayPal Marketing Solutions with a link to the settings so they can turn it on.

**Comment by kcppdevelopers** [comment]
@angelleye I have added admin notice and extra marketing verbiage / graphics like they have in theirs.

**Comment by angelleye** [comment]
Need to use "Marketing Solutions" in all places, not "Insights"

**Comment by angelleye** [comment]
JS integration from start to finish with no API call.

want to remove as much friction as possible

paypal integrated signup

"we don't want them to ever see the activate button"

since we're using the API call we don't need to show "Activate" at all.  We're getting our CID with API call, so we can just display.  

Double check API call to ensure we're enabling the options by default.  This way we don't need to click Manage Settings at all.

Second time through no need for check box to enable.

**Comment by angelleye** [comment]
I think we should also add some logic on our admin notice so that if this is already enabled then the notice doesn't show up.

**Comment by angelleye** [comment]
We also need to consider clearing any locally stored cid value if the API credentials are cleared or changed.  

For example, if you put in credentials for PayPal Account A, and then you enable Insights to generate a CID, but then you change the credentials to PayPal Account B, we need to be sure the CID is updated for this new account.

**Comment by kcppdevelopers** [comment]
@angelleye for me PayPal insights API it's not working on sandbox mode, and my PayPal live account insights not supported. please send me document when you get from PayPal.

**Comment by angelleye** [comment]
@kcppdevelopers We're going to leave the API call in place for now, but we do need to make another adjustment.

Instead of a check box to enable Insights, we need to create a button that looks just like the "Activate Insights" button that the JS code would load.  This would just be our button, though, that triggers our API call the same way we would do with the check box.  So we'd just click the Activate button, and the next thing the user would see is a message that "PayPal Marketing Solutions has been successfully activated for your site!"

Then once it's activated, the T&C stuff would go away and our button would be replaced with the JS code that would then load the "Manage Settings" button.  This way the user experience will be the same as the full JS integration for Activate and Manage, but we'd be using the API call to activate.

**Comment by angelleye** [comment]
Need to consider a way to "clear client ID".  Right now once you've activated there isn't any simple way to clear out the CID if you need to so that you can get the Activate button back again.

**Comment by angelleye** [comment]
@kcppdevelopers When I click the link in our admin notice it's no longer sending me directly to the Insights section of the EC settings the way it used to..??

**Comment by angelleye** [comment]
@kcppdevelopers The More Info button is opening up this:

![image](https://user-images.githubusercontent.com/629241/32653294-007974b8-c5cd-11e7-94aa-8aa3b9f7b3c3.png)

That window needs to be more of a "widescreen" instead of a square, more like what I provided in our wireframe/demo.

**Comment by angelleye** [comment]
@kcppdevelopers We also still need to consider how this is going to work smoothly with the regular settings in EC.  

For example, if I'm a brand new user setting things up in EC and I get all the way to bottom...then I click Activate without ever Saving the settings I've already configured, would I end up losing those settings I've adjusted?  If so, we need to either change it back to a check box so that it works smoothly with the rest of the EC settings in general, or we may move Marketing Solutions into a separate tab/page altogether.

Please let me know your thoughts on this.

**Comment by angelleye** [comment]
@kcppdevelopers Here is the link that should go with our "View Insights" button after the user activates it.

https://www.paypal.com/us/webapps/mpp/paypal-marketing-solutions

**Comment by kcppdevelopers** [comment]
> For example, if I'm a brand new user setting things up in EC and I get all the way to bottom...then I click Activate without ever Saving the settings I've already configured, would I end up losing those settings I've adjusted?

@angelleye When user click on Activate, we are not refreshing full page, you end up not losing those settings I've adjusted. we have also used ajax to save marketing value to database.

**Comment by angelleye** [comment]
There are some updates to the API we need to make.  To avoid duplicate setups of the same site with different CID values, they have added a check for this and will now return error info if it detects a duplicate it already exists.  It should return the same CID that was originally setup in this case.

Attached are samples of request/response with this functionality included.
[marketing-solutions-sample-api.zip](https://github.com/angelleye/paypal-woocommerce/files/1461130/marketing-solutions-sample-api.zip)

**Comment by angelleye** [comment]
Also, I think we're still missing something with the activation.  I still had to login and agree to activate it.  This should be skipped since we're using the API call.  Clicking the button should be able to simply activate everything behind the scenes via the API call so it simply activates for us without logging in and messing with toggles or anything.

**Comment by kcppdevelopers** [comment]
@angelleye  Please Note: we have used JS version for PayPal Marketing Solutions.

**Comment by angelleye** [comment]
@kcppdevelopers You were supposed to prepare it both ways..??  Remember I asked you to get the API version ready in a separate branch from JS version so we could launch either way..??

**Comment by angelleye** [comment]
@kcppdevelopers Maybe I need to switch to API version branch..??

**Comment by angelleye** [comment]
@kcppdevelopers We can use this graphic as the popup for the More Info button.

![paypal-for-woocommerce-marketing-solutions-more-info](https://user-images.githubusercontent.com/629241/32658530-fc9ba3b2-c5df-11e7-93b7-82574ac76062.jpg)

**Comment by angelleye** [comment]
The "partner_name" value included in the request is what gets displayed on the "Return to..." button:

![image](https://user-images.githubusercontent.com/629241/32661067-053a9b12-c5eb-11e7-83a2-190e09a48ec5.png)

So in this case we should use the site name pulled from WP settings.

**Comment by angelleye** [comment]
@kcppdevelopers Still some things we need to fix here:

* The BN code used is incorrect.  BN code is supposed to be `AngellEYE_SP_WooCommerce_MS`
* Partner Name should be the WP site name.

**Comment by angelleye** [comment]
@kcppdevelopers I pulled the dev branch, which only had a change in the main plugin file.  I uploaded that and ran another activation of Marketing Solutions, but I don't see any log file.  I also don't see any commit that includes logging..??

**Comment by angelleye** [comment]
@kcppdevelopers When I click the button to clear the Marketing Solution data (CID) it's actually wiping out the live API credentials, too.

**Comment by angelleye** [comment]
@kcppdevelopers For some reason now I am not getting the banners to show up at all, even when I clear site data via Chrome Console -> Application -> Storage -> Clear Site Data.  Not sure if this has something to do with the new params that were added in 580c225..??

**Comment by angelleye** [comment]
If we go back to JS integration we need to add the following:

![image](https://user-images.githubusercontent.com/629241/32763264-409bcfd8-c8c4-11e7-9dab-7e44d8e9faa6.png)

**Comment by angelleye** [comment]
The "analytics-id" that we're passing right now is a bogus/sample value.  We need to change this to use the PayPal Merchant Account ID, and then also add "-1" to the end of it.

We can use [GetPalDetails](https://developer.paypal.com/docs/classic/api/merchant/GetPalDetails_API_Operation_NVP/) to obtain the users Merchant Account ID.  That call is built into our library so it be a very simple call to add.  Then we just need the value we get back for PAL and append -1 to it (ie. BHXLBZQUN3LP8-1)

**Comment by angelleye** [comment]
* Fix API credentials from getting wiped out (remove API call from Enable check box, and instead make the API call when the Save button is pushed).
* Resize pop-up window to the same size as the More Info graphic (1327x833)
* Adjust log file so that the JSON request is formatted in a friendly / readable way.
